lib/propolis/src/tasks.rs

```diff
@@ -434,22 +434,27 @@ impl TaskGroup {
         guard.extend(tasks);
     }
 
-    /// Block until all held tasks have been joined, returning any resulting
-    /// [task::JoinError]s after doing so.
-    pub fn block_until_joined(&self) -> Option<Vec<task::JoinError>> {
-        let mut guard = self.0.lock().unwrap();
-        let workers = std::mem::replace(&mut *guard, Vec::new());
+    /// Waits until all the workers in this task group have completed.
+    ///
+    /// # Return value
+    ///
+    /// `None` if all the tasks completed successfully. `Some` if at least one
+    /// task failed; the wrapped value is a `Vec` of all of the returned errors.
+    pub async fn join_all(&self) -> Option<Vec<task::JoinError>> {
```

Comment:
Now that this is async, I think we could could conceivably replace it with a new implementation using [`tokio::task::JoinSet`](https://docs.rs/tokio/latest/tokio/task/join_set/struct.JoinSet.html) instead of a `Vec<JoinHandle<()>>`. This would be more efficient, as it wouldn't have to poll every `JoinHandle` in a loop. I'm not sure if this is hot enough that it matters if we do that or not, though, and it would be fine to do it in a follow-up.